### PR TITLE
Execution Cost Flamegraph

### PR DESCRIPTION
**Pull request by 0xOmarA**
## Summary

This PR adds a more detailed execution cost breakdown and also adds the ability for a flamegraph to be generated it.

faucet-free-xrd

(Save the above SVG locally to view it correctly)

## Details

This PR adds the ability for the costing module to collect more detailed information on the cost unit consumption during execution. This information is then surfaced in transaction receipts and can be graphed and represented as a flamegraph/flamechart. Note that in flamegraphs the events are ordered alphabetically and not in the order in which they occured.

Please note that this PR makes a breaking change to transaction receipts. I suggest that we keep this breaking change until we introduce the transaction receipt DTO.

### CODE REVIEW

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (0xOmarA) [inline radix-engine/src/transaction/transaction_receipt.rs:224]: Bringing this to the attention of the reviewers

**Review by dhedey — COMMENTED** [review]
Really nice output! :)

But before considering merging this, I think we need to be more careful about:
* What we're adding to our interfaces and the consequences. Indeed, I don't think we want to add lots of internal engine models to the transaction receipt, as it makes it much harder to change them without breaking things.
* Reducing performance impact for normal preview (which full nodes do a lot of)

Instead:
* Perhaps we may wish to flatten lots of things out into Strings on the way to the receipt
* OR consider pushing the flamegraph generation further into the engine, to improve perf and reduce the overhead.

If we want a more structured execution trace for other reasons, (which might be really cool, and was my original model for preview), we should start by considering _what_ the format of its DTO should be - rather than just exposing engine models on the receipt interface.

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (iamyulong) [reply radix-engine/src/transaction/transaction_receipt.rs:224]: Here is what I imagined how the DTO thing would work.

``` rust
VersionedTransactionReceipt {
    V1(TransactionReceiptV1) // DTO, identical to the current TransactionReceipt
    V2(TransactionReceiptV2) // another version if any
}
```

The above will be used by Wallet, Node and RET for compatibility. The `TransactionReceipt` in scrypto repo will be used for runtime only, e.g. being used by Node for mapping into receipt DTO or by Scrypto tests.

With this in mind, would it make sense to just mark the field as `skipped` so no compatibility-breaking change is introduced, before DTO is implemented.

Comment (dhedey) [reply radix-engine/src/transaction/transaction_receipt.rs:224]: @iamyulong this can't be used for the wallet, because you might have a wallet which only understands V1 talking to a node outputting V2. 

For now, we need to always return `VersionedTransactionReceipt::V1` from the API.

I suggest a DTO strategy for the preview will be preview-specific and be something with a permissive parser which supports additions (such as JSON or protocol buffers).

Comment (dhedey) [reply radix-engine/src/transaction/transaction_receipt.rs:224]: Using `#[sbor(skip)]` would also work! I forgot that that exists. I think it's possible the first concrete use case for it I've seen aha.

File: scrypto-test/src/environment/env.rs
Comment (0xOmarA) [reply scrypto-test/src/environment/env.rs:177]: This syntax is actually supported by both VS Code (and any IDE that uses Rust Analyzer) and also Rust Rover. Here is an issue that was opened in the RA repo adding support for this `region` syntax: https://github.com/rust-lang/rust-analyzer/issues/4217.

This currently has proper IDE support. As an example, in VS Code I get the option to collapse this particular region of code and it also shows up on the code navigator in VS Code.

image

Does the IDE support that this has address the readability concerns you had?

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (0xOmarA) [reply radix-engine/src/transaction/transaction_receipt.rs:224]: Thank you both for your input on this! I think we can just treat the way that the information is exposed in this PR as being temporary since I really just needed a way to get the information from the costing module and into the tests that consume the information. So, we can think of a better way to expose this information.

> We can't do this. For the time being, for wallet compatibility, we need TransactionReceiptV1 to be fixed.

Yup, agreed, this was more of a way to expose this information in some way until we agree on a proper way of exposing it.


> The DTO will need to be additional to the TransactionReceiptV1 until the wallet can migrate to the DTO and enough time has past that we have enough people using wallets with the DTO that we can remove the transaction receipt.

Yup, agreed. I share similar thoughts on this. Will outline some my thoughts on this at the end of this comment.

> With this in mind, would it make sense to just mark the field as skipped so no compatibility-breaking change is introduced, from persistence, encoding and decoding perspective, before DTO is implemented.

I like this! This would give us a way to do this before implementing the DTOs. I guess we would then need to make this field optional (not an issue, just more of a question)? I guess I'm not entirely sure what `#[sbor(skip)]` does to be able to reason about this.

---

I have a small proposal on how we can do the whole receipt versioning primarily with respect to the wallet.

1. We would continue to define `VersionedTransactionReceipt` as an enum of versions V1 all the way up to VN.
2. V(N+1) can always be downgraded to VN and VN can always be upgraded to V(N+1).
3. The core api and gateway api accept an additional argument in transaction preview of the version of receipt desired by the wallet.
4. Each wallet version knows the version of receipt it's compatible with and requests that from the gateway when doing a preview.
5. CoreAPI and Gateway would no longer return `VersionedTransactionReceipt` but would return `TransactionReceiptVX` where X is a version specified by the client in the request.

I see this as being a very simple way to go about this. Thoughts?

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (0xOmarA) [reply radix-engine/src/transaction/transaction_receipt.rs:99]: Yup, agreed, will change that.

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (0xOmarA) [reply radix-engine/src/transaction/transaction_receipt.rs:153]: > The names are rather long/noisy - I don't think we need Invocation or index here.
> I guess adding the index prevents the flamegraph merging common parents and makes the flamegraph into a timeline - BUT because it's not a PREFIX, it doesn't give the right ordering. So it's like the worst of both worlds.

The main reason why we have the index in there is to use it if further analysis is needed. As an example, we notice that some "read substate" is consuming too much cost units so we use the `index` in the flamegraph to identify which one it is in the `detailed_execution_cost_breakdown` and analyze it further. 

As you mentioned, in a flamegraph the order is not preserved so doing further analysis without having the index is going to be rather difficult. 

> Perhaps we can support a few modes:

I do not disagree, but would recommend first seeing what modes we _need_ during our analysis of the execution cost units and add them on need basis :)

File: radix-engine/src/system/system_modules/costing/costing_module.rs
Comment (0xOmarA) [reply radix-engine/src/system/system_modules/costing/costing_module.rs:163]: Agreed, seems like this would make sense. Will investigate making this into its own setting (I hope that it doesn't become a breaking change to transactions).

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (0xOmarA) [reply radix-engine/src/transaction/transaction_receipt.rs:79]: True, I think that the main issue is that there isn't quite an easy way to get a file path from a `std::fs::File` and the flamegraph library primarily works with paths.

I guess a sufficiently large random number might also be useable here instead of `file.txt`? Or perhaps even better we could just hash the `detailed_execution_cost_breakdown` and make that the name of the temporary file.

**Comment by 0xOmarA** [comment]
@dhedey Thank you for the review! I'm responding to your review comment in this comment.

> * What we're adding to our interfaces and the consequences. Indeed, I don't think we want to add lots of internal engine models to the transaction receipt, as it makes it much harder to change them without breaking things.

So, I thought about this a little bit more. I think that I would like _not_ to include this information in the transaction receipt. Instead, I want to propose that execution returns _two_ transactions receipts instead of one: `TransactionReceiptV1` and something like `DebugTransactionReceipt`.

The `TransactionReceiptV1` is the original transaction receipt that we've had for a while now, no changes to that, not even the detailed execution cost units breakdown that I added with this PR. The `DebugTransactionReceipt` (name is just an example for the time being) is another receipt returned that is not versioned and that would include information that we would never persist and would just include information that we want to expose to tests or for debugging. This model would not be versioned since it would not need versioning as its never persisted. We would never use this outside of tests. Additionally, this model can use as much of the internal engine types as it needs, again, since it's never persisted and not used anywhere other than tests.

This would address the backward compatibility issues we talked about and also any concerns we have related to the size of the preview receipts since adding this detailed execution cost information to the transaction receipt will make it massive.

What do you think? 

> * Reducing performance impact for normal preview (which full nodes do a lot of)

Totally agree, need to rework this bit. I'm now concerned about how much the size of the receipt increases with this information and about the performance impact. So, will rework this bit.

> Instead:
> * Perhaps we may wish to flatten lots of things out into Strings on the way to the receipt

At the current moment of time, I would recommend that we do not flatten them into strings and keep them this way so that we can extract as much information out of them as needed during the analysis. You'll notice that only a small amount of the information is used in the flamegraph and a large part goes unused, this is intentionally done so that we have this information when we need further analysis.

If we do the `DebugTransactionReceipt` idea, do you think that this would alleviate your concerns about the size? 

> * OR consider pushing the flamegraph generation further into the engine, to improve perf and reduce the overhead.

To understand, in this case the flamegraph would not be generated from the receipt, it would be generated on the fly (perhaps in the costing module) as the transaction is executed? 

> If we want a more structured execution trace for other reasons, (which might be really cool, and was my original model for preview), we should start by considering _what_ the format of its DTO should be - rather than just exposing engine models on the receipt interface.

Yup, I agree, I'm personally very much in favor of us not exposing this information at the time being since it's still not very mature and can very much change and instead we would surface it to tests in a `DebugTransactionReceipt` that accompanies any `TransactionReceipt` we get when executing tests.

File: radix-engine/src/system/system_modules/costing/costing_module.rs
Comment (0xOmarA) [reply radix-engine/src/system/system_modules/costing/costing_module.rs:164]: Sounds good, will do that.

File: radix-engine/src/system/actor.rs
Comment (0xOmarA) [reply radix-engine/src/system/actor.rs:5]: I think that this would be addressed by the `DebugTransactionReceipt` idea I mentioned [here](https://github.com/radixdlt/radixdlt-scrypto/pull/1834#issuecomment-2178658074). If we do it then we would no longer need to make those models SBOR encodable.

File: radix-engine/src/kernel/substate_io.rs
Comment (0xOmarA) [reply radix-engine/src/kernel/substate_io.rs:23]: I think that this would be addressed by the `DebugTransactionReceipt` idea I mentioned [here](https://github.com/radixdlt/radixdlt-scrypto/pull/1834#issuecomment-2178658074). If we do it then we would no longer need to make those models SBOR encodable.

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (0xOmarA) [reply radix-engine/src/transaction/transaction_receipt.rs:163]: Ah yeah it'd quite a few lines. Essentially _everything_ costed is a line; a read substate is a line, a write substate is a line, a drop node is a line, etc...

I think that we can potentially push this flamegraph generation into the engine but I think we would also want the detailed execution cost breakdown to be exposed to us somehow so we can use it for further analysis. 

I talk about a `DebugTransactionReceipt` idea in [this](https://github.com/radixdlt/radixdlt-scrypto/pull/1834#issuecomment-2178658074) comment here which I think is the right solution to some of the issues in this PR. Would love to get your opinion on that.

File: radix-engine/src/system/system_modules/costing/costing_module.rs
Comment (0xOmarA) [reply radix-engine/src/system/system_modules/costing/costing_module.rs:215]: Ah, seems like a depth of 0 might be better here. Will do that.

File: scrypto-test/src/environment/env.rs
Comment (dhedey) [reply scrypto-test/src/environment/env.rs:177]: Well this is just personal taste, but I don't really look at the side-view in VS Code much... And 
```
    // region:invocations
```
Is less nice to me in code. What about another compromise such as e.g.
```
    //==============
    // region: Invocations
    //==============


    //==============
    // endregion: Invocations
    // region: Kernel Access
    //==============
```

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (dhedey) [reply radix-engine/src/transaction/transaction_receipt.rs:79]: A quick google of "rust tempfile crate get path" gives me this https://docs.rs/tempfile/latest/tempfile/struct.TempPath.html - does that work?

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (iamyulong) [reply radix-engine/src/transaction/transaction_receipt.rs:224]: > @iamyulong this can't be used for the wallet, because you might have a wallet which only understands VersionedTransactionReceipt::V1 talking to a node outputting latest = VersionedTransactionReceipt::V2.

This is what typically wallet deprecation solves. In the future, you may not be able to convert V2 into a V2 for legacy clients.

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (0xOmarA) [reply radix-engine/src/transaction/transaction_receipt.rs:79]: Ah this is great! Will use that then!

File: scrypto-test/src/environment/env.rs
Comment (0xOmarA) [reply scrypto-test/src/environment/env.rs:177]: That sounds good to me! I've changed that.

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (0xOmarA) [reply radix-engine/src/transaction/transaction_receipt.rs]: Thank you all for this discussion! I will mark this as resolved since we've already discussed it on slack.

File: radix-engine/src/system/system_modules/costing/costing_module.rs
Comment (0xOmarA) [reply radix-engine/src/system/system_modules/costing/costing_module.rs]: I would recommend leaving this change for a later PR since it's quite a big change and the logic around the depth I think is somewhat simple and can easily be verified.

**Review by iamyulong — APPROVED** [review]
LGTM,  but ideally we merge after David's approval as he've given most of the comments. :-D

File: radix-engine/src/transaction/transaction_receipt.rs
Comment (0xOmarA) [reply radix-engine/src/transaction/transaction_receipt.rs:179]: > In Scrypto, we're charging tiny amounts - so how many costing calls do we get? One per cost batch? In an average dApp transaction, how many lines are we talking / how much memory are we talking out of interest?

About this bit, I've addressed it in the slack thread that we have for this PR.

> Also, Decimal is rather big (192 bits?) - I wonder if we should use e.g. u32 / cost units instead.

Yup, we're already using a `u32` of the cost units and not a `Decimal`.

File: radix-engine/src/system/system_callback.rs
Comment (dhedey) [inline radix-engine/src/system/system_callback.rs:870]: I think this was missed? We should resolve this TODO - either commenting why we can assume this, or handling it properly?

Seems reasonable to me that `init` sets `current_depth` to `0` though 👍

File: radix-engine/src/system/system_callback.rs
Comment (dhedey) [inline radix-engine/src/system/system_callback.rs:1132]: If we create a `costing_module.deconstruct()` which splits out the `detailed_cost_breakdown` from the `fee_reserve` etc then can we avoid the expensive clone here?

File: radix-engine/src/system/actor.rs
Comment (dhedey) [reply radix-engine/src/system/actor.rs:5]: So I was going to comment "we should remove Sbor from this and all these models"

But I guess this is so we can create the natural JSON output of the detailed trace for the tests?

I guess this is okay. We're happy to change these models because they are only used in tests :).

And we can catch it in future if someone tries to add this to an output model...

File: radix-engine/src/system/system_modules/costing/costing_module.rs
Comment (dhedey) [reply radix-engine/src/system/system_modules/costing/costing_module.rs]: Looks like you did fix this in the end, thanks :).

But yeah, I think it makes sense to change in this PR so we don't introduce something and then remove it which creates unneeded churn.

**Review by dhedey — APPROVED** [review]
Looks good, thanks. One minor comment about a TODO that we should probably fix, otherwise good to go.

File: radix-engine/src/system/system_callback.rs
Comment (0xOmarA) [reply radix-engine/src/system/system_callback.rs:1132]: Sounds good to me. In this case avoiding the clone didn't require adding a new method. The current context already owns the `CostingModule` so I can simply operate over the owned value instead of operating over a reference and this avoids the clone.

File: radix-engine/src/system/system_callback.rs
Comment (0xOmarA) [reply radix-engine/src/system/system_callback.rs:870]: Yup I missed this, I've removed the TODO comment and will add a comment explaining why we set it to zero.

**Pull request merged**
